src/NuGet.Packaging/PackagesConfigReader.cs

```diff
@@ -191,18 +202,25 @@ public IEnumerable<PackageReference> GetPackages()
             }
 
             // check if there are duplicate entries
-            var packageIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
-            var duplicates = new List<string>();
-            foreach (var package in packages)
+            IReadOnlyList<string> duplicates = null;
+
+            if (allowDuplicatePackageIds)
             {
-                if (packageIds.Contains(package.PackageIdentity.Id))
-                {
-                    duplicates.Add(package.PackageIdentity.Id);
-                }
-                else
-                {
-                    packageIds.Add(package.PackageIdentity.Id);
-                }
+                // Verify that no duplicate package identities
+                duplicates = packages.GroupBy(package => package.PackageIdentity, PackageIdentity.Comparer)
```

Comment:
+1. 
To test it, try with a packages.config with 200+ entries and see what the performance would be like.